New Edge sidebar PWA developer documentation (#2458)
MicrosoftDocs/edge-developer: microsoft-edge/progressive-web-apps-chromium/how-to/sidebar.md

---
title: Build Microsoft Edge sidebar PWAs
description: How to build your PWA for installation in the Microsoft Edge Sidebar.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: pwa
ms.date: 02/22/2023
---
# Build Microsoft Edge sidebar PWAs

![Experimental feature](../../media/experimental-tag.png)

Progressive Web Apps (PWAs) can opt-in to be pinned to the sidebar in Microsoft Edge.

The sidebar in Microsoft Edge allows users to easily access popular websites and utilities alongside their browser tabs. The content in the sidebar augments the user's primary task by enabling side-by-side browsing and minimizing the need to switch contexts between browser tabs.

By signaling intent to be pinned to the sidebar, your PWA gets the following benefits:

> * Your PWA is promoted in the Microsoft Edge sidebar store.

Reviewer: .. may be promoted. I imagine that eventually if we succeed, we will need to balance promotion with other recommendation criteria, e.g. user engagement with sites, etc.

* Users can still resize the sidebar to make it larger than your preferred width, but not smaller.

<!-- ====================================================================== -->
## Build a sidebar-only app

One of the most important benefits of building an app as a PWA is that, from one code base, your app adapts to all devices and operating systems, whatever their capabilities and screen sizes.

By using the `edge_side_panel` manifest member, you can also make your app installable as a standalone app, or pinned to the sidebar in Microsoft Edge.

However, if you prefer to build an app that can only be pinned to the sidebar but not installed as a standalone app, omit the `display` member from your web app manifest, or set its value to `browser`:

```json
{
  "name": "PWAmp music player",
  "lang": "en-US",
  "start_url": "/",
  "short_name": "PWAmp",
  "description": "A skinable music player app to play your favorite mp3 files",
  "display": "browser",
  "icons": [
    {
      "src": "./favicon-256.png",
      "sizes": "256x256"
    }
  ],
  "edge_side_panel": {
    "preferred_width": 480
  }
}
```

<!-- ====================================================================== -->

> ## Detect when your app runs in the sidebar

Reviewer: This could also be for sites.

Author: Not sure what to do here. This documentation (and the entire parent area) is focused on PWA apps only. So rephrasing to "Detect when your app or site runs in the sidebar" feels a bit out of place here.

Reviewer 2: Let's use "Detect usage in the sidebar" then

Reviewer 2: Detect usage in the sidebar

Detecting when your app runs in the sidebar can be useful to provide your users with the best possible experience when your app runs in this specific browser UI surface.

You can detect when your app runs in the sidebar by using User-Agent Client Hints. To learn more about User-Agent Client Hints, see [Detecting Microsoft Edge from your website](../../web-platform/user-agent-guidance.md).

To detect when your app runs in the sidebar, do one of the following:

> * On your web server, read the `Sec-CH-UA` HTTPS request header and look for the `Edge Side Panel` brand. For example:

Reviewer: As Hoch mentioned, client hint headers will not be in 112, but planned to be upcoming soon (Target 113). We could say, coming soon/next?

Author: Since we're waiting for feedback on the explainer anyway, maybe by the time we're ready to land the docs, the new CH headers will be available.
If they're not, I'll remove this part of the docs. It's better to document only what's available than to say some other pieces will come soon.

Reviewer 3: This is being enabled through `msAddEdgeSidePaneBrandUA` feature flag.

    
    ```https
    Sec-CH-UA: "Microsoft Edge";v="112", "Edge Side Panel";v="1", "Placeholder;Browser Brand";v="99"
    ```

* In the browser, use the `navigator.userAgentData` JavaScript API and read the value of the `brands` property. For example:

    ```javascript
    const brands = navigator.userAgentData.brands;
    const sidebarBrandInfo = brands.find(b => b.brand === "Edge Side Panel");
    if (sidebarBrandInfo) {
      console.log(sidebarBrandInfo); // { brand: "Edge Side Panel", version: "1" }
    } else {
      console.log("App is not running in the Microsoft Edge sidebar");
    }
    ```

You can also use the User Agent string to know whether your app runs in the sidebar or not.

> **Note**: we strongly recommend to use User Agent Client Hints instead. Using the User Agent string is an outdated mechanism for doing browser detection that has a long history of website compatibility issues.

Reviewer 4: ```suggestion
**Note**: we strongly recommend using User Agent Client Hints instead of the User Agent string. The User Agent string is an outdated mechanism for doing browser detection, and has website compatibility issues.
```
gentler yet still direct wording

* One variant when the app is used on a desktop device.
* Another variant when the app is used on a mobile device.

In this scenario, the desktop variant is used when your app is pinned to the sidebar. By default, apps in the sidebar receive the Microsoft Edge desktop User Agent Client Hint:

```https
Sec-CH-UA-Mobile: ?0
```

As explained in [Adapt your app to the minimum sidebar width](#adapt-your-app-to-the-minimum-sidebar-width), we recommend making your app's layout responsive so that it can adapt to multiple form factors using a single code base.

You can, however, use the `Edge Side Panel` User Agent Client Hint brand to reuse your app's mobile variant in the Microsoft Edge sidebar.

Follow these recommendations when using a mobile-variant of your app in the sidebar:

* Remove all "Open in App" messages instructing users to download your app from an app store.
* Test the accessibility and usability of your app with all input methods: mouse, keyboard, and touch. To learn about testing the accessibility of your app, see [Overview of accessibility testing using DevTools](../../devtools-guide-chromium/accessibility/accessibility-testing-in-devtools.md).

> <!-- ====================================================================== -->

Reviewer: Can we add a section to get feedback from developers on Sidebar features they'd like to add/see for PWAs? Not sure we have a GitHub for PWA specific developer feedback. Something that was discussed, but we'd like feedback on before reconsidering was the ability to specify a start_url in the edge_side_panel manifest section.

Reviewer: https://github.com/MicrosoftDocs/edge-developer/issues could be a reasonable place to direct Issues. Thanks for the pointer Hoch!

Author: We have the MSEdgeExplainers repo that can be used for user feedback about new features. I'd prefer if we could create a new explainer there to engage with people and collect their feedback rather than use the edge-developer docs repo.

Reviewer: Agreed. Adolf created this explainer now.: https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/SidePanel/explainer.md.

---
title: Build PWAs for the sidebar in Microsoft Edge

> description: How to build your PWA for installation in the Microsoft Edge Sidebar.

Reviewer 4: ```suggestion
description: How to build your PWA for installation in the Microsoft Edge sidebar.
```

author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: pwa
ms.date: 04/27/2023
---

> # Build Microsoft Edge sidebar PWAs

Reviewer 2: Build PWAs for the sidebar in Microsoft Edge

![Experimental feature](../../media/experimental-tag.png)

Progressive Web Apps (PWAs) can opt-in to be pinned to the sidebar in Microsoft Edge.

The sidebar in Microsoft Edge allows users to easily access popular websites and utilities alongside their browser tabs. The content in the sidebar augments the user's primary task by enabling side-by-side browsing and minimizing the need to switch contexts between browser tabs.

By signaling intent to be pinned to the sidebar, your PWA gets the following benefits:

* Your PWA can be promoted in the Microsoft Edge sidebar store.
* Your PWA can detect whether it is running in the Microsoft Edge sidebar or not.
* Your PWA defines its own style and layout to provide a user-friendly sidebar experience.

<!-- ====================================================================== -->
## Enable sidebar support of PWAs in Microsoft Edge

> Experimental support for pinning PWAs in the Microsoft Edge's sidebar starts with Microsoft Edge 112. Some parts of the feature are available starting with Microsoft Edge 114, such as the ability to detect if your PWA is running in the sidebar by using User Agent Client Hints.

Reviewer 2: Experimental support for PWAs that can target the sidebar in Microsoft Edge starts with ...

To locally enable PWA sidebar support in Microsoft Edge:

1. Download and install the Canary channel of Microsoft Edge from the [Microsoft Edge Insider website](https://www.microsoftedgeinsider.com/download).

> 1. Depending on your operating system, do either of the following to start Microsoft Edge with sidebar support for PWAs:

Reviewer 4: ```suggestion
1. Start Microsoft Edge with sidebar support for PWAs, as follows:
```

Reviewer 4: reworked list section b/c "either" combined with unordered lists was confusing to me

    * On Windows:
        

>         * Find the path where Microsoft Edge Canary is installed on your device. For example: `C:\Users\your_name\AppData\Local\Microsoft\Edge SxS\Application\msedge.exe`.
>         * Enter **cmd** in the Windows search text box and select the **Command Prompt** app.
>         * Enter `"C:\Users\your_name\AppData\Local\Microsoft\Edge SxS\Application\msedge.exe" --enable-features=msWebAppManifestSidePanel,msAddEdgeSidePaneBrandUA`.

Reviewer 4: ```suggestion
        1. Find the path where Microsoft Edge Canary is installed on your device. For example: `C:\Users\your_name\AppData\Local\Microsoft\Edge SxS\Application\msedge.exe`.
        1. Enter **cmd** in the Windows search text box, and then select the **Command Prompt** app.
        1. Enter `"C:\Users\your_name\AppData\Local\Microsoft\Edge SxS\Application\msedge.exe" --enable-features=msWebAppManifestSidePanel,msAddEdgeSidePaneBrandUA`.
```

    * On macOS:

>         * Find the path where Microsoft Edge Canary is installed on your device. For example: `/Applications/Microsoft Edge Canary.app`.
>         * Open the **Terminal** app.
>         * Enter `/Applications/Microsoft Edge Canary.app --enable-features=msWebAppManifestSidePanel,msAddEdgeSidePaneBrandUA`.

Reviewer 4: ```suggestion
        1. Find the path where Microsoft Edge Canary is installed on your device. For example: `/Applications/Microsoft Edge Canary.app`.
        1. Open the **Terminal** app.
        1. Enter `/Applications/Microsoft Edge Canary.app --enable-features=msWebAppManifestSidePanel,msAddEdgeSidePaneBrandUA`.
```

Reviewer 4: result according to VS Code's preview with Learn pack:
![image](https://user-images.githubusercontent.com/45407486/234967217-554b8f99-643c-4c49-8648-cbe5879b5dcf.png)

<!-- ====================================================================== -->
## Enable sidebar support in your PWA

> To make your PWA ready for pinning to the sidebar in Microsoft Edge, use the `edge_sidebar_panel` manifest member:

Reviewer 2: This is wrong. Should be edge_side_panel

Reviewer 2: Here and everywhere else

1. Make sure your PWA has a web app manifest file that contains at least the `name`, `short_name`, `description`, and `icons` members.

    To learn more, see [Use a web app manifest to integrate a Progressive Web App into the operating system](./web-app-manifests.md).

1. Add the `edge_side_panel` member to your web app manifest. For example:

    ```json
    {
      "name": "PWAmp music player",
      "lang": "en-US",
      "start_url": "/",
      "short_name": "PWAmp",

>       "description": "A skinable music player app to play your favorite mp3 files",

Reviewer 4: ```suggestion
      "description": "A skinnable music player app to play your favorite mp3 files",
```
https://www.bing.com/search?pglt=129&q=skinable -- 309k hits if include skinnable, drops to 65k if strictly skinable

      "display": "standalone",
      "icons": [
        {
          "src": "./favicon-256.png",
          "sizes": "256x256"
        }
      ],
      "edge_side_panel": {}
    }
    ```

<!-- ====================================================================== -->
## Adapt your app to the minimum sidebar width

> The sidebar in Microsoft Edge has a default minimum width of 376 pixels and can be resized by users. Therefore, your app's layout should support the 376 pixels minimum width and be responsive.

Reviewer 4: ```suggestion
The sidebar in Microsoft Edge has a default minimum width of 376 pixels and can be resized by users. Therefore, your app's layout should support the 376 pixels minimum width, and should be responsive.
```
separated the two thoughts

> While building your app, you can test if the app's layout is usable when narrow, and responsive, by using the **Device Emulation** tool in Microsoft Edge DevTools. To learn more, see [Verify that the webpage layout is usable when narrow](../../devtools-guide-chromium/accessibility/narrow.md).

Reviewer 4: ```suggestion
While building your app, you can test whether the app's layout is usable when narrow, and make sure the design is responsive at different widths, by using the **Device Emulation** tool in Microsoft Edge DevTools. To learn more, see [Verify that the webpage layout is usable when narrow](../../devtools-guide-chromium/accessibility/narrow.md).
```

If your app's layout can't support the 376 pixels minimum width, you can define your preferred width by using the `preferred_width` property in your web app manifest. For example:

```json
{
  "name": "PWAmp music player",
  "lang": "en-US",
  "start_url": "/",
  "short_name": "PWAmp",
  "description": "A skinable music player app to play your favorite mp3 files",
  "display": "standalone",
  "icons": [
    {
      "src": "./favicon-256.png",
      "sizes": "256x256"
    }
  ],
  "edge_side_panel": {
    "preferred_width": 480
  }
}
```

When you define a preferred width in your app's manifest, the following happens:

* When your app is opened in the sidebar, the sidebar is automatically resized to your preferred width.

> * Users can resize the sidebar to make it larger your preferred width, or smaller, up to the 376 pixels minimum width.

Reviewer 4: ```suggestion
* Users can resize the sidebar to make it larger than your preferred width, or smaller, up to the 376 pixels minimum width.
```

Reviewer 4: this sentence could be smoother but i'll look at others before trying to draft alternatives

<!-- ====================================================================== -->
## Build a sidebar-only app

One of the most important benefits of building an app as a PWA is that, from one code base, your app adapts to all devices and operating systems, whatever their capabilities and screen sizes.

> By using the `edge_side_panel` manifest member, you can also make your app installable as a standalone app, or pinned to the sidebar in Microsoft Edge.

Reviewer 4: ```suggestion
The `edge_side_panel` manifest member makes your app installable as a standalone app and enables your app to be pinned to the sidebar in Microsoft Edge.
```

> However, if you prefer to build an app that can only be pinned to the sidebar but not installed as a standalone app, omit the `display` member from your web app manifest, or set its value to `browser`:

Reviewer 4: ```suggestion
However, if you prefer to build an app that can only be pinned to the sidebar but cannot be not installed as a standalone app, omit the `display` member from your web app manifest, or set its value to `browser`:
```

```json
{
  "name": "PWAmp music player",
  "lang": "en-US",
  "start_url": "/",
  "short_name": "PWAmp",

>   "description": "A skinable music player app to play your favorite mp3 files",

Reviewer 4: ```suggestion
  "description": "A skinnable music player app to play your favorite mp3 files",
```

  "display": "browser",
  "icons": [
    {
      "src": "./favicon-256.png",
      "sizes": "256x256"
    }
  ],
  "edge_side_panel": {
    "preferred_width": 480
  }
}
```

<!-- ====================================================================== -->
## Detect usage in the sidebar

> Detecting when your app runs in the sidebar can be useful to provide your users with the best possible experience when your app runs in this specific browser UI surface.

Reviewer 4: ```suggestion
Detecting when your app runs in the sidebar can be useful to provide your users with the best possible experience when your app runs in this specific browser UI.
```
does adding "surface" give a net gain, or introduces as much unclarity as clarification?

Starting with Microsoft Edge 114, you can detect when your app runs in the sidebar by using User-Agent Client Hints. To learn more about User-Agent Client Hints, see [Detecting Microsoft Edge from your website](../../web-platform/user-agent-guidance.md).

To detect when your app runs in the sidebar, do one of the following:

* On your web server, read the `Sec-CH-UA` HTTPS request header and look for the `Edge Side Panel` brand. For example:
    
    ```https
    Sec-CH-UA: "Microsoft Edge";v="112", "Edge Side Panel";v="1", "Placeholder;Browser Brand";v="99"
    ```

* In the browser, use the `navigator.userAgentData` JavaScript API and read the value of the `brands` property. For example:

    ```javascript
    const brands = navigator.userAgentData.brands;
    const sidebarBrandInfo = brands.find(b => b.brand === "Edge Side Panel");
    if (sidebarBrandInfo) {
      console.log(sidebarBrandInfo); // { brand: "Edge Side Panel", version: "1" }
    } else {
      console.log("App is not running in the Microsoft Edge sidebar");
    }
    ```

> You can also use the User Agent string to know whether your app runs in the sidebar or not.

Reviewer 4: ```suggestion
To detect whether your app is running in the sidebar, you can also use the User Agent string.
```

**Note**: we strongly recommend using User Agent Client Hints instead of the User Agent string. The User Agent string is an outdated mechanism for doing browser detection, and has website compatibility issues.

> If you can't use User Agent Client Hints in your app, look for `EdgeSidePanel` in the User Agent string. Here are some examples of User Agent strings containing the `EdgeSidePanel` value:

Reviewer 3: ```suggestion
If you can't use User Agent Client Hints in your app, look for `Edge Side Panel` in the User Agent string. Here are some examples of User Agent strings containing the `Edge Side Panel` value:
```

```https

> User-Agent: ... (EdgeSidePanel)...

Reviewer 3: same below

User-Agent: ... (..., Edge Side Panel)...
User-Agent: ... (Edge Side Panel, ...)...
```

#### Relationship with your mobile app

> You might have created two variants of your app:
> 
> * One variant when the app is used on a desktop device.
> * Another variant when the app is used on a mobile device.

Reviewer 4: ```suggestion
Suppose you create a desktop variant of your app and a mobile variant of your app.
```
"You might have created ...." is awkward and it sounds like speculating about user (avoid).

> You might have created two variants of your app:
> 
> * One variant when the app is used on a desktop device.
> * Another variant when the app is used on a mobile device.
> 
> In this scenario, the desktop variant is used when your app is pinned to the sidebar. By default, apps in the sidebar that use the `edge_side_panel` manifest member receive the Microsoft Edge desktop User Agent Client Hint:

Reviewer 4: ```suggestion
Suppose you create a desktop variant of your app and a mobile variant of your app. In this scenario, the desktop variant is used when your app is pinned to the sidebar. By default, apps in the sidebar that use the `edge_side_panel` manifest member receive the Microsoft Edge desktop User Agent Client Hint:
```
"You might have created ...." is awkward and it sounds like speculating about user (avoid).
Simplified wording to make the concept seem simple and familiar, not requiring more precision.
removed blank line

You can, however, use the `Edge Side Panel` User Agent Client Hint brand to reuse your app's mobile variant in the Microsoft Edge sidebar.

Follow these recommendations when using a mobile-variant of your app in the sidebar:

* Remove all "Open in App" messages instructing users to download your app from an app store.
* Test the accessibility and usability of your app with all input methods: mouse, keyboard, and touch. To learn about testing the accessibility of your app, see [Overview of accessibility testing using DevTools](../../devtools-guide-chromium/accessibility/accessibility-testing-in-devtools.md).

<!-- ====================================================================== -->
## Demo app

PWAmp is a music player PWA demo application that can be pinned to the sidebar in Microsoft Edge. To test PWAmp as a sidebar app:

1. Enable sidebar support as described in [Enable sidebar support of PWAs in Microsoft Edge](#enable-sidebar-support-of-pwas-in-microsoft-edge).

> 1. Open Microsoft Edge and make sure the sidebar is displayed. If the sidebar is not displayed, go to `edge://settings/sidebar` and then enable **Always show sidebar**:

Reviewer 4: ```suggestion
1. Open Microsoft Edge and make sure the sidebar is displayed. If the sidebar is not displayed, go to `edge://settings/sidebar` and then turn on the **Always show sidebar** toggle:
```
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/t/toggle
https://learn.microsoft.com/en-us/style-guide/procedures-instructions/describing-interactions-with-ui - find 'toggle'
